Partial sync `bidirectional text` UA stylesheet rules as per Web Specification

Ahmad-S792 opened the pull request:
#### ea2a8afbd89e36c72f4c7785538e206f4778730a
<pre>
Partial sync `bidirectional text` UA stylesheet rules as per Web Specification

<a href="https://bugs.webkit.org/show_bug.cgi?id=284620">https://bugs.webkit.org/show_bug.cgi?id=284620</a>
<a href="https://rdar.apple.com/141426858">rdar://141426858</a>

Reviewed by Tim Nguyen.

This patch is to align WebKit with Web-Specification [1]:

[1] <a href="https://html.spec.whatwg.org/multipage/rendering.html#bidi-rendering">https://html.spec.whatwg.org/multipage/rendering.html#bidi-rendering</a>

It syncs partially bidirectional text user agent stylesheet rules to ensure that
respective HTML elements have respect direction based as per standard.

The WPT test changes are synced as per:

Upstream commit: <a href="https://github.com/web-platform-tests/wpt/commit/e45da1637c6ae670bdc45929905506fca4f6d0bd">https://github.com/web-platform-tests/wpt/commit/e45da1637c6ae670bdc45929905506fca4f6d0bd</a>

* Source/WebCore/css/html.css:
(hr):
(address, blockquote, center, div, figure, figcaption, footer, form, header, hr,):
(bdi, output): Deleted.
* Source/WebCore/editing/EditingStyle.cpp:
(EditingStyle::textDirectionForSelection): textDirection could be `embed` or `isolate` so assert both
* Source/WebCore/editing/EditingStyle.h:

Test Updates (API &amp; Layout):
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/bidi-rendering/unicode-bidi-ua-rules-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/flow-content-0/search-styles-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/flow-content-0/search-styles-iso-8859-8-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-expected.txt: Progression
* LayoutTests/fast/css/all-keyword-direction.html: Added `3px` tolerance for Windows Platform
* LayoutTests/fast/css/all-keyword-direction-expected.html: Rebaselined
* LayoutTests/fast/css/default-bidi-css-rules.html: Updated test for new behavior
* LayoutTests/fast/css/default-bidi-css-rules-expected.txt: Rebaselined
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt: Rebaseline
* LayoutTests/fast/css/unicode-bidi-computed-value-expected.txt: Rebaselined
* LayoutTests/fast/css/unicode-bidi-computed-value.html: Fixed test based on change done by Blink
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/unicode-bidi-parsing-001.html: Synced from WPT
* LayoutTests/inspector/css/shadow-scoped-style-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/forms/form-in-malformed-markup-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/forms/form-in-malformed-markup-expected.txt: Rebaselined
* LayoutTests/platform/gtk/fast/forms/form-in-malformed-markup-expected.txt: Rebaselined
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEditActions.mm:
(ModifyTextWritingDirection): Update Test

Canonical link: <a href="https://commits.webkit.org/287842@main">https://commits.webkit.org/287842@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c99ae2365d03ce3ab9ea9bf06ac4ca5d9bad9ab8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81059 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/584 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35002 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85588 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32045 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83169 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/602 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8389 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63276 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21041 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84128 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/364 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73777 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43574 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/261 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27937 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30503 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71801 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28492 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87023 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8289 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5857 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71579 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8466 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70815 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17628 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14860 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13784 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8250 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13773 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8087 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11607 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9895 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on LayoutTests/fast/css/default-bidi-css-rules-expected.txt:
```diff
@@ -1,15 +1,15 @@
 
 PASS styleOf("div", {}).direction is "ltr"
-FAIL styleOf("div", {}).unicodeBidi should be isolate. Was normal.
+PASS styleOf("div", {}).unicodeBidi is "isolate"
```
Do you know if all of these are covered by WPT? Maybe we could begin deleting some of these `fast/css` test cases where they are covered by WPT. (Note: This is an idea for a follow-up patch, not for today's patch).

Ahmad-S792 replied:
Yes - I also think now that we have WPT, we should get rid of it.

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/html/rendering/bidi-rendering/unicode-bidi-ua-rules-expected.txt:
```diff
@@ -1,52 +1,52 @@
+PASS UA stylesheet rule for unicode-bidi, for <figcaption>
+PASS UA stylesheet rule for unicode-bidi, for <footer>
+PASS UA stylesheet rule for unicode-bidi, for <form>
+PASS UA stylesheet rule for unicode-bidi, for <header>
```
Lovely to see so many FAIL->PASS changes!

brentfulgham commented on LayoutTests/platform/gtk/fast/forms/form-in-malformed-markup-expected.txt:
```diff
@@ -4,7 +4,7 @@ layer at (0,0) size 800x600
     RenderBody {BODY} at (8,8) size 784x584
       RenderBlock (anonymous) at (0,0) size 784x0
-        RenderInline {B} at (0,0) size 0x0
+        RenderInline {B} at (0,-14) size 0x17
```
I'm not sure if this is right. Maybe @nt1m could confirm?

Ahmad-S792 replied:
I can make test to use old behavior by adding `* { unicode-bidi: normal }` like other tests. I didn't done it because I didn't notice rendering difference which is worse off like other cases.

brentfulgham commented on Source/WebCore/css/html.css:
```diff
@@ -1406,8 +1405,11 @@ iframe {
+address, blockquote, center, div, figure, figcaption, footer, form, header, hr,
+legend, listing, main, p, plaintext, pre, summary, xmp, article, aside, h1, h2,
+h3, h4, h5, h6, hgroup, nav, section, search, table, caption, colgroup, col,
+thead, tbody, tfoot, tr, td, th, dir, dd, dl, dt, menu, ol, ul, li, bdi, output {
```
How do we know that this is the set that should have this style attached, and none others? Is that documented somewhere?

Ahmad-S792 replied:
It is from web specification here - https://html.spec.whatwg.org/multipage/rendering.html#bidi-rendering

brentfulgham replied:
Oh, excellent. I would add that in your ChangeLog to indicate where that came from.

Ahmad-S792 replied:
It is already in commit. :-)

brentfulgham reviewed (commented):
I think this looks like a good change, but would like @nt1m to do the final review.

fantasai commented on LayoutTests/fast/css/all-keyword-direction.html:
```diff
@@ -12,6 +12,10 @@
+
+* {
+  unicode-bidi: normal;
+}
```
What is this for? AFAICT it should have no effect on this file.

Ahmad-S792 replied:
It pushes `.` to front of text instead of end without this. I can either update test to retain old behavior with this change or can update `-expected.html` file and we match all other browsers (Chrome Canary 133 and Firefox Nightly 135) without this test change with new behavior. (JSFiddle - https://jsfiddle.net/j3gdqh0p/)

They also push `.` at front / start instead of end. So now it is up to us whether we update test or `-expected.html` file.

Ahmad-S792 replied:
Updated `-expected.html` file instead of test.

fantasai commented on LayoutTests/fast/events/contextmenu-actions-in-selected-text.html:
```diff
@@ -8,6 +8,10 @@
+
+* {
+    unicode-bidi: normal;
+}
```
Same question: why do we need this rule?

Ahmad-S792 replied:
Without this change, we hit assertion in Editing Code, where we assume `embed` while it should be `isolate or embed`, which I have to investigate and I wanted to do it as separate change rather than merging in it.

nt1m replied:
We should make sure the assertion in editing code does not trigger.

Ahmad-S792 replied:
I know but do you want me to tackle this in this patch?

nt1m replied:
Yes, we should not introduce new regressions/assertions. Makes it harder to debug websites when they are hit.

Ahmad-S792 replied:
Removed test changes and fixed assert - locally ran with no issue.

Ahmad-S792 committed:
Partial sync `bidirectional text` UA stylesheet rules as per Web Specification

https://bugs.webkit.org/show_bug.cgi?id=284620
rdar://141426858

Reviewed by Tim Nguyen.

This patch is to align WebKit with Web-Specification [1]:

[1] https://html.spec.whatwg.org/multipage/rendering.html#bidi-rendering

It syncs partially bidirectional text user agent stylesheet rules to ensure that
respective HTML elements have respect direction based as per standard.

The WPT test changes are synced as per:

Upstream commit: https://github.com/web-platform-tests/wpt/commit/e45da1637c6ae670bdc45929905506fca4f6d0bd

* Source/WebCore/css/html.css:
(hr):
(address, blockquote, center, div, figure, figcaption, footer, form, header, hr,):
(bdi, output): Deleted.
* Source/WebCore/editing/EditingStyle.cpp:
(EditingStyle::textDirectionForSelection): textDirection could be `embed` or `isolate` so assert both
* Source/WebCore/editing/EditingStyle.h:

Test Updates (API & Layout):
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/bidi-rendering/unicode-bidi-ua-rules-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/flow-content-0/search-styles-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/flow-content-0/search-styles-iso-8859-8-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-expected.txt: Progression
* LayoutTests/fast/css/all-keyword-direction.html: Added `3px` tolerance for Windows Platform
* LayoutTests/fast/css/all-keyword-direction-expected.html: Rebaselined
* LayoutTests/fast/css/default-bidi-css-rules.html: Updated test for new behavior
* LayoutTests/fast/css/default-bidi-css-rules-expected.txt: Rebaselined
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt: Rebaseline
* LayoutTests/fast/css/unicode-bidi-computed-value-expected.txt: Rebaselined
* LayoutTests/fast/css/unicode-bidi-computed-value.html: Fixed test based on change done by Blink
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/unicode-bidi-parsing-001.html: Synced from WPT
* LayoutTests/inspector/css/shadow-scoped-style-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/forms/form-in-malformed-markup-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/forms/form-in-malformed-markup-expected.txt: Rebaselined
* LayoutTests/platform/gtk/fast/forms/form-in-malformed-markup-expected.txt: Rebaselined
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEditActions.mm:
(ModifyTextWritingDirection): Update Test

Canonical link: https://commits.webkit.org/287842@main
